TOOLS/lua/autocrop.lua

```diff
@@ -1,114 +1,344 @@
+
+auto: bool - Whether to automatically apply crop at the start of playback.
+    If you don't want to crop automatically, set it to false or
+    add "script-opts-append=autocrop-auto=no" into mpv.conf.
+
+auto_delay: seconds - Delay before starting crop in auto mode.
+    You can try to increase this value to avoid dark scene or fade in at beginning.
+    Automatic cropping will not occur if the value is larger than the remaining playback time.
+
+detect_limit: number[0-255] - Black threshold for cropdetect.
+    Smaller values will generally result in less cropping.
+    See limit of https://ffmpeg.org/ffmpeg-filters.html#cropdetect
+
+detect_round: number[2^n] -  The value which the width/height should be divisible by.
+    Smaller values ​​have better detection accuracy.
+    If you have problems with other filters, you can try to set it to 4 or 16.
+    See round of https://ffmpeg.org/ffmpeg-filters.html#cropdetect
+
+detect_min_ratio: number[0.0-1.0] - The ratio of the minimum clip size to the original.
+    If the picture is over cropped or under cropped, try adjusting this value.
+
+detect_seconds: seconds - How long to gather cropdetect data.
+    Increasing this may be desirable to allow cropdetect more time to collect data.
+--]]
+
+
 require "mp.msg"
+require 'mp.options'
 
-script_name = mp.get_script_name()
-cropdetect_label = string.format("%s-cropdetect", script_name)
-crop_label = string.format("%s-crop", script_name)
 
--- number of seconds to gather cropdetect data
-detect_seconds = tonumber(mp.get_opt(string.format("%s.detect_seconds", script_name)))
-if not detect_seconds then
+local options = {
+    auto = true,
+    auto_delay = 4,
+    detect_limit = 24,
+    detect_round = 2,
+    detect_min_ratio = 0.5,
     detect_seconds = 1
-end
+}
+read_options(options)
 
-function del_filter_if_present(label)
-    -- necessary because mp.command('vf del @label:filter') raises an
-    -- error if the filter doesn't exist
-    local vfs = mp.get_property_native("vf")
-    for i,vf in pairs(vfs) do
-        if vf["label"] == label then
-            table.remove(vfs, i)
-            mp.set_property_native("vf", vfs)
+
+local label_prefix = mp.get_script_name()
+local labels = {
+    crop = string.format("%s-crop", label_prefix),
+    cropdetect = string.format("%s-cropdetect", label_prefix)
+}
+
+timers = {
+    auto_delay = nil,
+    detect_crop = nil
+}
+
+
+function is_filter_present(label)
+    local filters = mp.get_property_native("vf")
+    for index, filter in pairs(filters) do
+        if filter["label"] == label then
             return true
         end
     end
     return false
 end
 
-function autocrop_start()
-    -- exit if cropdetection is already in progress
-    if timer then
-        mp.msg.warn("already cropdetecting!")
-        return
+
+function is_time_enough(seconds)
+
+    -- Plus 1 second for deviation.
+    local time_needed = seconds + 1
+
+
+    -- For still images
+    if not is_playing_video() then
```

Comment:
I'm not sure if this makes sense now. This function should really be called `is_album_art` and if the answer is yes, you still need sufficient time to let the crop detect filter run.

Reply:
In fact, this is an ugly trick let users can manually crop still images.
But in my opinion, still images have no concept of time, so there is no problem of not enough time.

Reply:
The issue is that if you have a song playing with albumart the cropdetect filter will still need to run for 1 second before applying any filter.